### PR TITLE
Select last or specific boundary set for merging process

### PR DESCRIPTION
**Pull request by FranckLecuyer**
PR that must be merged before (migration merge orchestrator to Postgres) : 
https://github.com/gridsuite/merge-orchestrator-server/pull/50
https://github.com/gridsuite/deployment/pull/145

All following PR must be merged together : 
https://github.com/gridsuite/cgmes-boundary-server/pull/15
https://github.com/gridsuite/cgmes-boundary-import-job/pull/3
https://github.com/gridsuite/merge-orchestrator-server/pull/51
https://github.com/gridsuite/merge-notification-server/pull/16
https://github.com/gridsuite/gridmerge-app/pull/67

### CODE REVIEW

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmEntity.java [hunk 1]
```diff
@@ -7,37 +7,57 @@
...
+    @Column(name = "replacingDate")
     private LocalDateTime replacingDate;
 
+    @Column(name = "replacingBusinessProcess")
     private String replacingBusinessProcess;
 
+    @Column(name = "boundary")
+    @ElementCollection
+    @CollectionTable(foreignKey = @ForeignKey(name = "igmEntity_boundaries_fk", foreignKeyDefinition = "FOREIGN KEY (IgmEntity_date, IgmEntity_processUuid, IgmEntity_tso) REFERENCES merge_igm ON DELETE CASCADE"), indexes = {@Index(name = "igmEntity_boundaries_idx", columnList = "IgmEntity_processUuid, IgmEntity_date, IgmEntity_tso")})
     private List<UUID> boundaries;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmEntity.java:56]: Two attributes EQ_UUID et TP_UUID is maybe more optimized (size and acces database)

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java [hunk 2]
```diff
@@ -19,25 +20,18 @@
  * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
  */
 @Repository
-public interface IgmRepository extends CassandraRepository<IgmEntity, IgmEntityKey> {
+public interface IgmRepository extends JpaRepository<IgmEntity, IgmEntityKey> {
 
-    @Query("SELECT * FROM merge_igm WHERE processUuid = :processUuid")
+    @Query(value = "SELECT igm from #{#entityName} as igm where igm.key.processUuid = :processUuid")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:25]: Really  necessary ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java [hunk 3]
```diff
@@ -19,25 +20,18 @@
...
     List<IgmEntity> findByProcessUuid(UUID processUuid);
 
-    @Query("SELECT * FROM merge_igm WHERE processUuid = :processUuid AND date = :date")
+    @Query(value = "SELECT igm from #{#entityName} as igm where igm.key.processUuid = :processUuid and igm.key.date = :date")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:28]: Really necessary ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java [hunk 4]
```diff
@@ -19,25 +20,18 @@
...
     List<IgmEntity> findByProcessUuidAndDate(UUID processUuid, LocalDateTime date);
 
-    @Query("DELETE FROM merge_igm WHERE processUuid = :processUuid")
+    @Modifying
+    @Query(value = "DELETE from #{#entityName} as igm where igm.key.processUuid = :processUuid")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:32]: Really necessary ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java [hunk 5]
```diff
@@ -19,25 +20,18 @@
...
     void deleteByProcessUuid(UUID processUuid);
 
-    @Query("SELECT * FROM merge_igm WHERE processUuid = :processUuid AND date >= :minDate AND date <= :maxDate")
-    List<IgmEntity> findByProcessUuidAndInterval(UUID processUuid, LocalDateTime minDate, LocalDateTime maxDate);
-
-    @Query("SELECT * FROM merge_igm WHERE processUuid = :processUuid AND date = :date AND tso = :tso")
+    @Query(value = "SELECT igm from #{#entityName} as igm where igm.key.processUuid = :processUuid and igm.key.date = :date and igm.key.tso = :tso")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:35]: Really necessary ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java [hunk 6]
```diff
@@ -20,14 +21,31 @@
...
 @Repository
-public interface MergeRepository extends CassandraRepository<MergeEntity, MergeEntityKey> {
+public interface MergeRepository extends JpaRepository<MergeEntity, MergeEntityKey> {
 
-    @Query("SELECT * FROM merge WHERE processUuid = :processUuid")
-    List<MergeEntity> findByProcessUuid(@Param("processUuid") UUID processUuid);
-
-    @Query("DELETE FROM merge WHERE processUuid = :processUuid")
+    @Modifying
+    @Query(value = "DELETE from #{#entityName} as merge where merge.key.processUuid = :processUuid")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:27]: Really necessary ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmEntity.java (hunk 1 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmEntity.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java [hunk 7]
```diff
@@ -76,15 +77,29 @@ public CgmesBoundaryService(RestTemplate restTemplate) {
...
+        try {
+            ResponseEntity<Map<String, String>> responseEntity = cgmesBoundaryServerRest.exchange(uri, HttpMethod.GET, HttpEntity.EMPTY, new ParameterizedTypeReference<>() {
+            });
+            Map<String, String> body = responseEntity.getBody();
+            if (body != null) {
+                return Optional.of(new BoundaryContent(body.get(ID_KEY), body.get(FILE_NAME_KEY), body.get(BOUNDARY_KEY)));
+            } else {
+                LOGGER.error("Error searching boundary with id {} : body is null {}", boundaryId, responseEntity);
+            }
+        } catch (HttpStatusCodeException e) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java:94]: Done

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java [hunk 8]
```diff
@@ -204,8 +228,8 @@ private void merge(ProcessConfig processConfig, ZonedDateTime dateTime, String d
                 }
             });
             if (!igmEntities.stream().skip(1).allMatch(e -> e.getBoundaries() != null && boundariesIgms.equals(new TreeSet<>(e.getBoundaries())))) {
-                LOGGER.warn("IGMs for merge process {} {} at {} have been imported with different last boundaries !!!", processConfig.getProcess(), processConfig.getBusinessProcess(), date);
-            } else {
+                LOGGER.warn("IGMs for merge process {} {} at {} have been imported with different boundaries !!!", processConfig.getProcess(), processConfig.getBusinessProcess(), date);
+            } else if (processConfig.isUseLastBoundarySet()) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java]: Done

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 2 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:25]: primary key is composite key

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 3 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:28]: primary key is composite key

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:32]: primary key is composite key

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:35]: primary key is composite key

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:27]: primary key is composite key

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 2 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:25]: Yes and Spring Data JPA use also the query derivd by name fot the composite keys :
`@Repository
public interface IgmRepository extends JpaRepository<IgmEntity, IgmEntityKey> {
    
    List<IgmEntity> findByKeyProcessUuid(UUID processUuid);
    
    List<IgmEntity> findByKeyProcessUuidAndKeyDate(UUID processUuid, LocalDateTime date);

    @Modifying
    void deleteByKeyProcessUuid(UUID processUuid);
    
    Optional<IgmEntity> findByKeyProcessUuidAndKeyDateAndKeyTso(UUID processUuid, LocalDateTime date, String tso);
}`

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 3 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:28]: See above

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 4 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:32]: See above

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 5 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java:35]: See above

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 6 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:27]: See above

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java [hunk 9]
```diff
@@ -20,14 +21,31 @@
  * @author Franck Lecuyer <franck.lecuyer at rte-france.com
  */
 @Repository
-public interface MergeRepository extends CassandraRepository<MergeEntity, MergeEntityKey> {
+public interface MergeRepository extends JpaRepository<MergeEntity, MergeEntityKey> {
 
-    @Query("SELECT * FROM merge WHERE processUuid = :processUuid")
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:25]: Delete the specific query and use the standard Spring Data JPA with methods names

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 6 above)
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:27]: Delete the specific query and use the standard Spring Data JPA with methods names

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 2 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 3 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/IgmRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 9 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java:25]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

File: src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/repositories/MergeRepository.java]: Done in PR for migration to PostGres : https://github.com/gridsuite/merge-orchestrator-server/pull/50

**Review by SlimaneAmar — APPROVED** [review]

**Review by SlimaneAmar — APPROVED** [review]

File: src/main/java/org/gridsuite/merge/orchestrator/server/dto/BoundaryInfo.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+package org.gridsuite.merge.orchestrator.server.dto;
+
+import io.swagger.annotations.ApiModel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import java.time.LocalDateTime;
+
+/**
+ * @author Franck Lecuyer <etienne.homer at rte-france.com>
```
Comment (niconoir) [inline src/main/java/org/gridsuite/merge/orchestrator/server/dto/BoundaryInfo.java:16]: Change email address

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java [hunk 11]
```diff
@@ -340,6 +365,13 @@ private static Merge toMerge(MergeRepository.MergeIgm merge) {
 
         // handle each missing or invalid igms
         for (String tso : missingOrInvalidTsos) {
+            List<BoundaryContent> configBoundaries = getProcessConfigBoundaries(config);
+            if (configBoundaries.isEmpty()) {
+                var errorMessage = PROCESS + " " + config.getProcess() + " (" + config.getBusinessProcess() + ") : EQ and/or TP boundary not available !!";
+                mergeEventService.addErrorEvent(config.getProcessUuid(), config.getBusinessProcess(), errorMessage);
+                return replacingIGMs;
+            }
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java]: Done

File: src/main/java/org/gridsuite/merge/orchestrator/server/dto/BoundaryInfo.java (hunk 10 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/dto/BoundaryInfo.java]: Corrected

**Pull request merged**